woocommerce/woocommerce-ios
# Replace concrete usage with protocol: `ProductFormDataModel`, `ProductFormActionsFactoryProtocol`, and `TaxClassRequestable`

Prerequisite for #2085 

This PR abstracts parts of concrete usage with protocol in the product form and its related classes, so that `ProductVariation` can be supported for these classes in a future subtask. These changes should be pretty straightforward. I left the more complicated protocol that abstracts the current `ProductFormViewModel` for a future subtask. In addition, I left the `Product` protocol replacement in price and inventory settings for later since these 2 screens have UI changes between `Product` and `ProductVariation` (p91TBi-33Q-p2).

## Changes

- Created `ProductFormActionsFactoryProtocol` protocol that the existing `ProductFormActionsFactory` conforms to, and updated its existing usage
- Created `ProductFormDataModel` protocol that abstracts the `Product` usage in product form. It contains all the necessary fields for `ProductFormViewController` and its navigated view controllers. Replaced `Product` with `ProductFormDataModel` in product form related classes where replacement is straightforward
- Created `TaxClassRequestable` protocol that abstracts the `Product` usage in `TaxClassAction.requestMissingTaxClasses` within price settings. Updated existing usage in price settings classes
- Removed unused `product: Product` associated value in `ProductFormSection.PrimaryFieldRow.images` enum case
- Removed redundant extension helper `isShippingEnabled` in Yosemite layer and moved its unit tests from Yosemite to app layer.

## Testing

No changes in app behavior are expected, feel free to test for regression with the following steps:

- Go to the Products tab
- Tap on a simple product
- Tap on the description row, and make sure the editing behavior stays the same
- Go back to the product details
- Tap on the price settings row, and make sure the editing behavior stays the same especially on the tax class selector
- Go back to the product details
- Tap on the shipping settings row, and make sure the editing behavior stays the same especially on the shipping class selector
- Go back to the product details
- Tap on the "+" cell in the images header, and make sure image editing behavior stays the same

Update release notes:
